# Automatización para los procesos de análisis de PS (Conacyt)

> Por: Jesús Octavio García Flores, 
> octavio5545@ciencias.unam.mx,  2022

In [1]:
import pandas as pd
import numpy as np

In [2]:
url = "Datase_general_ps.csv"
datos = pd.read_csv(url)

C:\Users\octav\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3165: DtypeWarning: Columns (9) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


In [4]:
datos.head()

CLAVE_FONDO DESCRIPCION_CORTA_FONDO DESCRIPCION_LARGA_FONDO  \
0       DEMOS                   DEMOS      Fondo Demostración   
1       DEMOS                   DEMOS      Fondo Demostración   
2       DEMOS                   DEMOS      Fondo Demostración   
3       DEMOS                   DEMOS      Fondo Demostración   
4       DEMOS                   DEMOS      Fondo Demostración   

  CLAVE_CONVOCATORIA DESCRIPCION_LARGA_CONVOCATORIA FECHA_INICIO_CONVOCATORIA  \
0         CONVODEMOS      Convocatoria Demostración   2011-07-10 00:00:00.000   
1         CONVODEMOS      Convocatoria Demostración   2010-01-22 00:00:00.000   
2         CONVODEMOS      Convocatoria Demostración   2011-01-22 00:00:00.000   
3         CONVODEMOS      Convocatoria Demostración   2006-07-05 00:00:00.000   
4         CONVODEMOS      Convocatoria Demostración   2010-01-12 00:00:00.000   

    FECHA_FIN_CONVOCATORIA CLAVE_MODALIDAD                    MODALIDAD  \
0  2018-12-31 00:00:00.000              A1  Invest. Científica Aplicada   
1  2010-02-28 00:00:00.000              A1  Invest. Científica Aplicada   
2  2012-02-29 00:00:00.000              A1  Invest. Científica Aplicada   
3  2015-09-30 00:00:00.000              A1  Invest. Científica Aplicada   
4  2012-07-10 00:00:00.000              A1  Invest. Científica Aplicada   

    CLAVE_PROYECTO  
0  000000000146702  
1  000000000146702  
2  000000000146702  
3  000000000146702  
4  000000000146702

In [4]:
df = pd.DataFrame(datos)

In [35]:
# Nota: Tarda en ejecutarse esta celda.
df.nunique(axis=1)

0           9
1           9
2           9
3           9
4           9
           ..
1976462     9
1976463     9
1976464    10
1976465    10
1976466    10
Length: 1976467, dtype: int64

In [6]:
df.nunique()

CLAVE_FONDO                          189
DESCRIPCION_CORTA_FONDO              173
DESCRIPCION_LARGA_FONDO              180
CLAVE_CONVOCATORIA                  1640
DESCRIPCION_LARGA_CONVOCATORIA      1419
FECHA_INICIO_CONVOCATORIA           1258
FECHA_FIN_CONVOCATORIA              1244
CLAVE_MODALIDAD                      344
MODALIDAD                             24
CLAVE_PROYECTO                    110336
dtype: int64

_

### Agrupamiento por CLAVE_FONDO

In [6]:
clave_fondo = df.groupby(by=['CLAVE_FONDO'])

In [7]:
clave_fondo.head()

CLAVE_FONDO DESCRIPCION_CORTA_FONDO         DESCRIPCION_LARGA_FONDO  \
0             DEMOS                   DEMOS              Fondo Demostración   
1             DEMOS                   DEMOS              Fondo Demostración   
2             DEMOS                   DEMOS              Fondo Demostración   
3             DEMOS                   DEMOS              Fondo Demostración   
4             DEMOS                   DEMOS              Fondo Demostración   
...             ...                     ...                             ...   
1975215       FOP16                     FCF             Ciencia de Frontera   
1975216       FOP16                     FCF             Ciencia de Frontera   
1976464       FOP17                    UAUC  UNIDAD ACCESO UNIVERSAL CONOCI   
1976465       FOP17                    UAUC  UNIDAD ACCESO UNIVERSAL CONOCI   
1976466       FOP17                    UAUC  UNIDAD ACCESO UNIVERSAL CONOCI   

        CLAVE_CONVOCATORIA DESCRIPCION_LARGA_CONVOCATORIA  \
0               CONVODEMOS      Convocatoria Demostración   
1               CONVODEMOS      Convocatoria Demostración   
2               CONVODEMOS      Convocatoria Demostración   
3               CONVODEMOS      Convocatoria Demostración   
4               CONVODEMOS      Convocatoria Demostración   
...                    ...                            ...   
1975215      FOP16-2021-01     Paradigmas y Controversias   
1975216      FOP16-2021-01     Paradigmas y Controversias   
1976464      FOP17-2022-01  Proyectos de acceso universal   
1976465      FOP17-2022-01  Proyectos de acceso universal   
1976466      FOP17-2022-01  Proyectos de acceso universal   

        FECHA_INICIO_CONVOCATORIA   FECHA_FIN_CONVOCATORIA CLAVE_MODALIDAD  \
0         2011-07-10 00:00:00.000  2018-12-31 00:00:00.000              A1   
1         2010-01-22 00:00:00.000  2010-02-28 00:00:00.000              A1   
2         2011-01-22 00:00:00.000  2012-02-29 00:00:00.000              A1   
3         2006-07-05 00:00:00.000  2015-09-30 00:00:00.000              A1   
4         2010-01-12 00:00:00.000  2012-07-10 00:00:00.000              A1   
...                           ...                      ...             ...   
1975215   2021-11-29 00:00:00.000  2021-12-10 00:00:00.000              P1   
1975216   2021-11-29 00:00:00.000  2021-12-10 00:00:00.000              P1   
1976464   2022-03-28 00:00:00.000  2022-04-29 00:00:00.000               A   
1976465   2022-03-28 00:00:00.000  2022-04-29 00:00:00.000               A   
1976466   2022-03-28 00:00:00.000  2022-04-29 00:00:00.000               A   

                           MODALIDAD   CLAVE_PROYECTO  
0        Invest. Científica Aplicada  000000000146702  
1        Invest. Científica Aplicada  000000000146702  
2        Invest. Científica Aplicada  000000000146702  
3        Invest. Científica Aplicada  000000000146702  
4        Invest. Científica Aplicada  000000000146702  
...                              ...              ...  
1975215                          NaN           321051  
1975216                          NaN           320960  
1976464                  Modalidad A           321333  
1976465                  Modalidad A           321332  
1976466                  Modalidad A           321359  

[915 rows x 10 columns]

In [10]:
clave_fondo_unicos = clave_fondo.nunique()
clave_fondo_unicos

DESCRIPCION_CORTA_FONDO  DESCRIPCION_LARGA_FONDO  \
CLAVE_FONDO                                                     
AVANC                              1                        1   
C0001                              1                        1   
C0002                              1                        1   
C0003                              1                        1   
C0004                              1                        1   
...                              ...                      ...   
U0003                              1                        1   
U0004                              1                        1   
U0005                              1                        1   
U0006                              1                        1   
U0007                              1                        1   

             CLAVE_CONVOCATORIA  DESCRIPCION_LARGA_CONVOCATORIA  \
CLAVE_FONDO                                                       
AVANC                         5                               5   
C0001                         2                               2   
C0002                         1                               1   
C0003                         7                               7   
C0004                         3                               3   
...                         ...                             ...   
U0003                         9                               7   
U0004                         2                               1   
U0005                         4                               4   
U0006                         1                               1   
U0007                         1                               1   

             FECHA_INICIO_CONVOCATORIA  FECHA_FIN_CONVOCATORIA  \
CLAVE_FONDO                                                      
AVANC                                6                       5   
C0001                                2                       2   
C0002                                1                       1   
C0003                                4                       5   
C0004                                3                       3   
...                                ...                     ...   
U0003                                9                       7   
U0004                                2                       2   
U0005                                4                       4   
U0006                                1                       1   
U0007                                1                       1   

             CLAVE_MODALIDAD  MODALIDAD  CLAVE_PROYECTO  
CLAVE_FONDO                                              
AVANC                      5          0             657  
C0001                      1          1             553  
C0002                      2          0             501  
C0003                      7          3           10438  
C0004                      4          2             264  
...                      ...        ...             ...  
U0003                     16          2             370  
U0004                      1          0              20  
U0005                      1          0              40  
U0006                      1          0              21  
U0007                      1          0               4  

[189 rows x 9 columns]

Las columnas que deberían contener el mismo número de datos únicos son: 

- CLAVE_FONDO
- DESCRIPCION_CORTA_FONDO
- DESCRIPCION_LARGA_FONDO


Y también:

- CLAVE_CONVOCATORIA
- DESCRIPCION_LARGA_CONVOCATORIA	


### CLAVES_FONDO QUE TIENEN MÁS DE UNA DESCRIPCIÓN CORTA O LARGA.

In [27]:
# dist_descripciones = []

# for i, dato in enumerate(clave_fondo_unicos.DESCRIPCION_CORTA_FONDO):
#     indice_dato = clave_fondo_unicos.DESCRIPCION_CORTA_FONDO.to_list().index(dato)
#     print("índice_dato=", indice_dato)
#     if dato == clave_fondo_unicos.DESCRIPCION_LARGA_FONDO[i]:
#         dist_descripciones.append(indice_dato)

descrip_distintas = clave_fondo_unicos[clave_fondo_unicos.DESCRIPCION_CORTA_FONDO != 1]


In [29]:
len(descrip_distintas)

0

***No hay CLAVES_FONDO que contengan descripciones distintas (ni cortas, ni largas).


-

#### CUENTA LAS CLAVES_FONDO CUYAS COLUMNAS  "CLAVE_CONVOCATORIA"  TIENEN DESCRIPCION LARGA DE CONVOCATORIA DISTNTAS:

In [31]:
clave_descrip = clave_fondo_unicos[clave_fondo_unicos.CLAVE_CONVOCATORIA != clave_fondo_unicos.DESCRIPCION_LARGA_CONVOCATORIA]

In [32]:
clave_descrip

DESCRIPCION_CORTA_FONDO  DESCRIPCION_LARGA_FONDO  \
CLAVE_FONDO                                                     
C0013                              1                        1   
E0006                              1                        1   
F0001                              1                        1   
F0002                              1                        1   
F0004                              1                        1   
F0005                              1                        1   
F0007                              1                        1   
F0008                              1                        1   
FOP02                              1                        1   
I0002                              1                        1   
I0010                              1                        1   
I0015                              1                        1   
I0016                              1                        1   
I0022                              1                        1   
M0014                              1                        1   
M0021                              1                        1   
S0016                              1                        1   
S0018                              1                        1   
U0001                              1                        1   
U0003                              1                        1   
U0004                              1                        1   

             CLAVE_CONVOCATORIA  DESCRIPCION_LARGA_CONVOCATORIA  \
CLAVE_FONDO                                                       
C0013                        19                              16   
E0006                         5                               2   
F0001                         8                               4   
F0002                        11                              10   
F0004                         2                               1   
F0005                        17                               7   
F0007                         5                               3   
F0008                         8                               5   
FOP02                         5                               4   
I0002                        13                              11   
I0010                        17                              11   
I0015                         7                               1   
I0016                        17                              16   
I0022                         2                               1   
M0014                        36                              35   
M0021                        42                              41   
S0016                        10                               9   
S0018                        19                              18   
U0001                         2                               1   
U0003                         9                               7   
U0004                         2                               1   

             FECHA_INICIO_CONVOCATORIA  FECHA_FIN_CONVOCATORIA  \
CLAVE_FONDO                                                      
C0013                               22                      21   
E0006                                5                       5   
F0001                               18                      17   
F0002                               22                      21   
F0004                                2                       2   
F0005                               21                      21   
F0007                                6                       6   
F0008                               12                      11   
FOP02                                5                       5   
I0002                               20                      23   
I0010                               16                      17   
I0015                                6                       6   
I0016                               17             

In [33]:
len(clave_descrip)

21

In [38]:
# CF_DLC_dif :  CLAVE_FONDO con Descripciones Largas de Convocatoria Diferentes.

CF_DLC_dif = clave_descrip.index
print(f"\nLas siguientes claves fondo contienen diferentes descipciones largas: \n\n {CF_DLC_dif.values}")



Las siguientes claves fondo contienen diferentes descipciones largas: 

 ['C0013' 'E0006' 'F0001' 'F0002' 'F0004' 'F0005' 'F0007' 'F0008' 'FOP02'
 'I0002' 'I0010' 'I0015' 'I0016' 'I0022' 'M0014' 'M0021' 'S0016' 'S0018'
 'U0001' 'U0003' 'U0004']



#df[df.CLAVE_FONDO == CF_DLC_dif]
df_copia = df.copy()

for i, dato in enumerate(df_copia.CLAVE_FONDO):
    print(dato)
    if dato not in CF_DLC_dif:
        print(f"\n{dato}  no está en índices\n")
        df_copia.CLAVE_FONDO.drop(i)

In [41]:
df[df['CLAVE_FONDO'] == 'AVANC']

CLAVE_FONDO DESCRIPCION_CORTA_FONDO DESCRIPCION_LARGA_FONDO  \
24995        AVANC                  AVANCE         Programa AVANCE   
24996        AVANC                  AVANCE         Programa AVANCE   
24997        AVANC                  AVANCE         Programa AVANCE   
24998        AVANC                  AVANCE         Programa AVANCE   
24999        AVANC                  AVANCE         Programa AVANCE   
...            ...                     ...                     ...   
664283       AVANC                  AVANCE         Programa AVANCE   
664284       AVANC                  AVANCE         Programa AVANCE   
664285       AVANC                  AVANCE         Programa AVANCE   
664286       AVANC                  AVANCE         Programa AVANCE   
664287       AVANC                  AVANCE         Programa AVANCE   

       CLAVE_CONVOCATORIA DESCRIPCION_LARGA_CONVOCATORIA  \
24995     CONV_AVANCE2003       Convocatoria AVANCE 2003   
24996     CONV_AVANCE2003       Convocatoria AVANCE 2003   
24997     CONV_AVANCE2003       Convocatoria AVANCE 2003   
24998     CONV_AVANCE2003       Convocatoria AVANCE 2003   
24999     CONV_AVANCE2003       Convocatoria AVANCE 2003   
...                   ...                            ...   
664283    CONV_AVANCE2009       Convocatoria AVANCE 2009   
664284    CONV_AVANCE2009       Convocatoria AVANCE 2009   
664285    CONV_AVANCE2009       Convocatoria AVANCE 2009   
664286    CONV_AVANCE2009       Convocatoria AVANCE 2009   
664287    CONV_AVANCE2009       Convocatoria AVANCE 2009   

       FECHA_INICIO_CONVOCATORIA   FECHA_FIN_CONVOCATORIA CLAVE_MODALIDAD  \
24995    2009-11-17 00:00:00.000  2010-01-30 00:00:00.000           AVANC   
24996    2008-09-10 00:00:00.000  2009-11-16 00:00:00.000           AVANC   
24997    2007-07-19 00:00:00.000                      NaN           AVANC   
24998    2005-01-01 00:00:00.000  2007-12-31 00:00:00.000           AVANC   
24999    2003-01-01 00:00:00.000  2004-10-17 00:00:00.000           AVANC   
...                          ...                      ...             ...   
664283   2008-09-10 00:00:00.000  2009-11-16 00:00:00.000              EC   
664284   2007-07-19 00:00:00.000                      NaN              EC   
664285   2005-01-01 00:00:00.000  2007-12-31 00:00:00.000              EC   
664286   2003-01-01 00:00:00.000  2004-10-17 00:00:00.000              EC   
664287   2006-09-03 00:00:00.000  2006-09-04 00:00:00.000              EC   

       MODALIDAD CLAVE_PROYECTO  
24995        NaN     0000005624  
24996        NaN     0000005624  
24997        NaN     0000005624  
24998        NaN     0000005624  
24999        NaN     0000005624  
...          ...            ...  
664283       NaN         127812  
664284       NaN         127812  
664285       NaN         127812  
664286       NaN         127812  
664287       NaN         127812  

[3942 rows x 10 columns]

In [25]:
df.drop_duplicates("CLAVE_FONDO")\
.groupby(by=["DESCRIPCION_CORTA_FONDO"])\
.filter(lambda f1: f1.shape[0]>1) \
.sort_values("DESCRIPCION_CORTA_FONDO", ascending=True)

CLAVE_FONDO DESCRIPCION_CORTA_FONDO         DESCRIPCION_LARGA_FONDO  \
1975210       FOP15                CIBIOGEM                        CIBIOGEM   
727649        I0016                CIBIOGEM                        CIBIOGEM   
1836698       I0029                    DACI        Centros de Investigación   
1418890       U0004                    DACI                 PRODECyT - DACI   
1790874       I0031                    DADC           Desarrollo Científico   
941150        E0006                    DADC        APOYO INSTITUCIONAL DADC   
1549863       C0015                    DADC                 FONCICYT - DADC   
1355046       U0003                    DADC                 PRODECyT - DADC   
1016369       EE004                   DADER    APOYOS INSTITUCIONALES DADER   
1572701       U0002                   DADER                PRODECyT - DADER   
1809244       I0033                   DADTI  Desarrollo Tecnológico e Innov   
1755887       C0022                   DADTI  Desarrollo Tecnológico e Innov   
1464473       C0014                   DADTI                FONCICYT - DADTI   
1439070       U0005                   DADTI                PRODECyT - DADTI   
1533077       U0006                    DAPB                 PRODECyT - DAPB   
1644089       C0016                    DAPB                 FONCICYT - DAPB   
1599551       I0028                    DAPE         DIRECCIÓN DE PLANEACIÓN   
1758730       I0035                    DAPE         Planeación y Evaluación   
1425054       C0013                FONCICYT                      FONCICYT 2   
408410        C0002                FONCICYT                        FONCICYT   
1947255       FOP01                     FPC    FORDECYT CIENCIA con Pp F003   
1947512       FOP06                     FPC                    COMUNICACION   
1188774       R0003                REVISTAS  Indice Revistas de Divulgación   
617364        R0001                REVISTAS      Revistas Cientificas y Tec   
1168319       E0011              SENER-HIDR             SENER-HIDROCARBUROS   
554430        S0018              SENER-HIDR             SENER-HIDROCARBUROS   
764097        I0017                     SEP             Fondo SEP - CONACYT   
59716         I0003                     SEP             Fondo SEP - CONACYT   

        CLAVE_CONVOCATORIA DESCRIPCION_LARGA_CONVOCATORIA  \
1975210      FOP15-2022-01                  CIBIOGEM 2022   
727649           DIAG_MAIZ           Diagnostico del maiz   
1836698      I0029-2019-04   APOYOS DIRECTOS DACI 2019-04   
1418890       U0004-2014-1                  PRODECyT-DACI   
1790874      I0031-2019-01      APOYOS DIRECTOS DADC 2019   
941150       E0006-2011-01               A ESPECIAL  DADC   
1549863      C0015-2014-01                       ANR 2014   
1355046       U0003-2014-1                  PRODECyT-DADC   
1016369        ANEXO42-R38                           2012   
1572701       U0002-2015-2                 PRODECyT-DADER   
1809244      I0033_2019-01      Apoyos Directo DADTI 2019   
1755887      C0022-2017-01         Convocatoria CoBI 2017   
1464473      C0014-2014-01                 BPIFRANCE 2014   
1439070       U0005-2014-1               PRODECyT - DADTI   
1533077      U0006-2015-01                PRODECyT - DAPB   
1644089      C0016-2016-02                      ECOS 2016   
1599551      I0028-2015-02           CONVOCATORIA 2015-02   
1758730      I0035-2017-01   Repositorios Institucionales   
1425054      C0013-2014-01                   PIRE-CONACYT   
408410       C0002-2008-01     Convocatoria C0002-2008-01   
1947255      FOP01-2020-01              VIVIENDA ADECUADA   
1947512      FOP06-2020-01     Establecimiento de una Red   
1188774      R0003-2013-01           CONVOCATORIA 2013-01   
617364       R0001-2009-01                  R0001-2009-01   
1168319      E0011-2013-01           CONVOCATORIA 2013-01   
554430       S0018-01-2009           Convocatoria 2009-01   
764097          SEB-2009-1                   SEB  2009-01   
59716   

In [26]:
primary_keys_df = df[['CLAVE_FONDO', 'CLAVE_CONVOCATORIA', 'CLAVE_PROYECTO']]

In [27]:
primary_keys_df.head()

CLAVE_FONDO CLAVE_CONVOCATORIA   CLAVE_PROYECTO
0       DEMOS         CONVODEMOS  000000000146702
1       DEMOS         CONVODEMOS  000000000146702
2       DEMOS         CONVODEMOS  000000000146702
3       DEMOS         CONVODEMOS  000000000146702
4       DEMOS         CONVODEMOS  000000000146702

In [29]:
primary_keys_df.nunique()

CLAVE_FONDO              189
CLAVE_CONVOCATORIA      1640
CLAVE_PROYECTO        110336
dtype: int64

In [33]:
claves_fondo = df.CLAVE_FONDO.unique()
claves_fondo = pd.DataFrame(claves_fondo, columns=['CLAVE_FONDO'])
claves_fondo

CLAVE_FONDO
0         DEMOS
1         M0023
2         M0020
3         M0004
4         M0030
..          ...
184       FOP03
185       FOP14
186       FOP15
187       FOP16
188       FOP17

[189 rows x 1 columns]

In [34]:
claves_convocatoria = df.CLAVE_CONVOCATORIA.unique()
claves_convocatoria = pd.DataFrame(claves_convocatoria, columns=['CLAVE_CONVOCATORIA'])
claves_convocatoria

CLAVE_CONVOCATORIA
0            CONVODEMOS
1          M0023-2005-4
2          M0020-2006-1
3          M0004-2005-5
4          M0030-2005-1
...                 ...
1635      FOP04-2022-01
1636      FOP07-2022-01
1637      FOP07-2022-02
1638      FOP10-2022-01
1639      FOP17-2022-01

[1640 rows x 1 columns]

In [35]:
claves_proyectos = df.CLAVE_PROYECTO.unique()
claves_proyectos = pd.DataFrame(claves_proyectos, columns=['CLAVE_PROYECTO'])
claves_proyectos

CLAVE_PROYECTO
0       000000000146702
1       000000000146701
2       000000000118218
3       000000000109116
4       000000000107894
...                 ...
110331           321321
110332           321370
110333           321333
110334           321332
110335           321359

[110336 rows x 1 columns]

In [25]:
def filtro_repetidos(df, drop_colName, group_colName):
    df_filtrado = df.drop_duplicates(drop_colName)\
    .groupby(by=[group_colName])\
    .filter(lambda f1: f1.shape[0]>1) \
    .sort_values(group_colName, ascending=True)
    
    return df_filtrado

In [40]:
# df.drop_duplicates("CLAVE_CONVOCATORIA")\
# .groupby(by=["CLAVE_PROYECTO"])\
# .filter(lambda f1: f1.shape[0]>1) \
# .sort_values("CLAVE_PROYECTO", ascending=True)

filtro_repetidos(df, 'CLAVE_CONVOCATORIA', 'CLAVE_PROYECTO')

Empty DataFrame
Columns: [CLAVE_FONDO, DESCRIPCION_CORTA_FONDO, DESCRIPCION_LARGA_FONDO, CLAVE_CONVOCATORIA, DESCRIPCION_LARGA_CONVOCATORIA, FECHA_INICIO_CONVOCATORIA, FECHA_FIN_CONVOCATORIA, CLAVE_MODALIDAD, MODALIDAD, CLAVE_PROYECTO]
Index: []

In [41]:
filtro_repetidos(df, 'CLAVE_PROYECTO', 'CLAVE_CONVOCATORIA')

CLAVE_FONDO DESCRIPCION_CORTA_FONDO DESCRIPCION_LARGA_FONDO  \
141859        I0003                     SEP     Fondo SEP - CONACYT   
141419        I0003                     SEP     Fondo SEP - CONACYT   
141399        I0003                     SEP     Fondo SEP - CONACYT   
141379        I0003                     SEP     Fondo SEP - CONACYT   
141359        I0003                     SEP     Fondo SEP - CONACYT   
...             ...                     ...                     ...   
1637413       M0024               ZACATECAS         FOMIX ZACATECAS   
1741974       M0024               ZACATECAS         FOMIX ZACATECAS   
1741942       M0024               ZACATECAS         FOMIX ZACATECAS   
1741910       M0024               ZACATECAS         FOMIX ZACATECAS   
1742006       M0024               ZACATECAS         FOMIX ZACATECAS   

        CLAVE_CONVOCATORIA DESCRIPCION_LARGA_CONVOCATORIA  \
141859      AFSEDF-2006-01                 AFSEDF-2006-01   
141419      AFSEDF-2006-01                 AFSEDF-2006-01   
141399      AFSEDF-2006-01                 AFSEDF-2006-01   
141379      AFSEDF-2006-01                 AFSEDF-2006-01   
141359      AFSEDF-2006-01                 AFSEDF-2006-01   
...                    ...                            ...   
1637413        ZAC-2016-02              ZACATECAS-2016-02   
1741974        ZAC-2017-01              ZACATECAS-2017-01   
1741942        ZAC-2017-01              ZACATECAS-2017-01   
1741910        ZAC-2017-01              ZACATECAS-2017-01   
1742006        ZAC-2017-01              ZACATECAS-2017-01   

        FECHA_INICIO_CONVOCATORIA   FECHA_FIN_CONVOCATORIA CLAVE_MODALIDAD  \
141859    2009-10-19 00:00:00.000  2010-03-15 00:00:00.000               S   
141419    2009-10-19 00:00:00.000  2010-03-15 00:00:00.000               S   
141399    2009-10-19 00:00:00.000  2010-03-15 00:00:00.000               S   
141379    2009-10-19 00:00:00.000  2010-03-15 00:00:00.000               S   
141359    2009-10-19 00:00:00.000  2010-03-15 00:00:00.000               S   
...                           ...                      ...             ...   
1637413   2017-10-10 00:00:00.000  2017-10-27 00:00:00.000              A1   
1741974   2017-10-10 00:00:00.000  2017-10-27 00:00:00.000              C1   
1741942   2017-10-10 00:00:00.000  2017-10-27 00:00:00.000              C1   
1741910   2017-10-10 00:00:00.000  2017-10-27 00:00:00.000              C1   
1742006   2017-10-10 00:00:00.000  2017-10-27 00:00:00.000              C1   

                             MODALIDAD CLAVE_PROYECTO  
141859                             NaN          56485  
141419                             NaN          57386  
141399                             NaN          57407  
141379                             NaN          57949  
141359                             NaN          53456  
...                                ...            ...  
1637413    Invest. Científica Aplicada         273509  
1741974  Formación de Recursos Humanos         291839  
1741942  Formación de Recursos Humanos         291779  
1741910  Formación de Recursos Humanos         292048  
1742006  Formación de Recursos Humanos         291787  

[110159 rows x 10 columns]

**La llave única para el dataset PS incluye los campos CLAVE_FONDO, CLAVE_CONVOCATORIA y CLAVE_PROYECTO, porque los proyectos tienen asociadas una única convocatoria y estas a su vez un fondo**

##### Analizar la cantidad mínima y máxima de proyectos por fondo y convocatoria:

In [5]:
por_fondos = df.groupby(by='CLAVE_FONDO').count()

In [10]:
por_fondos.CLAVE_PROYECTO.describe()

count       189.000000
mean      10457.497354
std       25845.882706
min           1.000000
25%          75.000000
50%         792.000000
75%        6744.000000
max      195280.000000
Name: CLAVE_PROYECTO, dtype: float64

.

Con base en los fondos, el mínimo de proyectos que involucra un fondo es de **1** proyectos, mientras que el máximo es de **195280** proyectos.

.

In [7]:
por_convocat = df.groupby(by='CLAVE_CONVOCATORIA').count()
por_convocat.CLAVE_PROYECTO.describe()

count     1640.000000
mean      1205.162805
std       3124.452333
min          1.000000
25%         90.000000
50%        290.000000
75%        937.000000
max      48516.000000
Name: CLAVE_PROYECTO, dtype: float64

.

Con base en las convocatorias, el mínimo de proyectos que involucra una convocatoria es de **1** proyectos, mientras que el máximo es de **48516** proyectos.

.

In [8]:

for col in df.columns.values:
    nulls = df[col].isnull().sum()
    if nulls != 0:
        print(f"\033[1m--> Contiene valores nulos: \t {str(col)} \033[0m")
    else:
        print(f"No contiene valores nulos: \t {str(col)}")

No contiene valores nulos: 	 CLAVE_FONDO
No contiene valores nulos: 	 DESCRIPCION_CORTA_FONDO
No contiene valores nulos: 	 DESCRIPCION_LARGA_FONDO
No contiene valores nulos: 	 CLAVE_CONVOCATORIA
No contiene valores nulos: 	 DESCRIPCION_LARGA_CONVOCATORIA
--> Contiene valores nulos: 	 FECHA_INICIO_CONVOCATORIA 
--> Contiene valores nulos: 	 FECHA_FIN_CONVOCATORIA 
No contiene valores nulos: 	 CLAVE_MODALIDAD
--> Contiene valores nulos: 	 MODALIDAD 
No contiene valores nulos: 	 CLAVE_PROYECTO


.

#### No exiten valores nulos en: CLAVE_FONDO, CLAVE_CONVOCATORIA ni CLAVE_PROYECTO. Esto significa que todos los proyectos tienen asociado un FONDO del cual proviene.

.

In [ ]:
df[df['CLAVE_PROYECTO'].countnt]

In [11]:
por_fondos.CLAVE_PROYECTO==1

CLAVE_FONDO
AVANC    False
C0001    False
C0002    False
C0003    False
C0004    False
         ...  
U0003    False
U0004    False
U0005    False
U0006    False
U0007    False
Name: CLAVE_PROYECTO, Length: 189, dtype: bool

In [19]:
CP_ = df.groupby(['CLAVE_PROYECTO'])

In [23]:
CP_.head()

CLAVE_FONDO DESCRIPCION_CORTA_FONDO         DESCRIPCION_LARGA_FONDO  \
0             DEMOS                   DEMOS              Fondo Demostración   
1             DEMOS                   DEMOS              Fondo Demostración   
2             DEMOS                   DEMOS              Fondo Demostración   
3             DEMOS                   DEMOS              Fondo Demostración   
4             DEMOS                   DEMOS              Fondo Demostración   
...             ...                     ...                             ...   
1976462       FOP10                    FVNI  Vinculación E Nacional Interna   
1976463       FOP10                    FVNI  Vinculación E Nacional Interna   
1976464       FOP17                    UAUC  UNIDAD ACCESO UNIVERSAL CONOCI   
1976465       FOP17                    UAUC  UNIDAD ACCESO UNIVERSAL CONOCI   
1976466       FOP17                    UAUC  UNIDAD ACCESO UNIVERSAL CONOCI   

        CLAVE_CONVOCATORIA DESCRIPCION_LARGA_CONVOCATORIA  \
0               CONVODEMOS      Convocatoria Demostración   
1               CONVODEMOS      Convocatoria Demostración   
2               CONVODEMOS      Convocatoria Demostración   
3               CONVODEMOS      Convocatoria Demostración   
4               CONVODEMOS      Convocatoria Demostración   
...                    ...                            ...   
1976462      FOP10-2022-01         Encargo de Estado 2022   
1976463      FOP10-2022-01         Encargo de Estado 2022   
1976464      FOP17-2022-01  Proyectos de acceso universal   
1976465      FOP17-2022-01  Proyectos de acceso universal   
1976466      FOP17-2022-01  Proyectos de acceso universal   

        FECHA_INICIO_CONVOCATORIA   FECHA_FIN_CONVOCATORIA CLAVE_MODALIDAD  \
0         2011-07-10 00:00:00.000  2018-12-31 00:00:00.000              A1   
1         2010-01-22 00:00:00.000  2010-02-28 00:00:00.000              A1   
2         2011-01-22 00:00:00.000  2012-02-29 00:00:00.000              A1   
3         2006-07-05 00:00:00.000  2015-09-30 00:00:00.000              A1   
4         2010-01-12 00:00:00.000  2012-07-10 00:00:00.000              A1   
...                           ...                      ...             ...   
1976462   2021-10-07 00:00:00.000  2021-10-12 00:00:00.000             BB1   
1976463   2021-10-07 00:00:00.000  2021-10-12 00:00:00.000             BB1   
1976464   2022-03-28 00:00:00.000  2022-04-29 00:00:00.000               A   
1976465   2022-03-28 00:00:00.000  2022-04-29 00:00:00.000               A   
1976466   2022-03-28 00:00:00.000  2022-04-29 00:00:00.000               A   

                           MODALIDAD   CLAVE_PROYECTO  
0        Invest. Científica Aplicada  000000000146702  
1        Invest. Científica Aplicada  000000000146702  
2        Invest. Científica Aplicada  000000000146702  
3        Invest. Científica Aplicada  000000000146702  
4        Invest. Científica Aplicada  000000000146702  
...                              ...              ...  
1976462                          NaN           321370  
1976463                          NaN           321370  
1976464                  Modalidad A           321333  
1976465                  Modalidad A           321332  
1976466                  Modalidad A           321359  

[505431 rows x 10 columns]

In [17]:
CP_['CL']  # Ejecutará error porque no existe el campo CL.

KeyError: 'Column not found: False'

In [26]:
filtro_repetidos(df, 'CLAVE_CONVOCATORIA', 'FECHA_INICIO_CONVOCATORIA')

CLAVE_FONDO DESCRIPCION_CORTA_FONDO         DESCRIPCION_LARGA_FONDO  \
835971        M0007              Guanajuato                FOMIX GUANAJUATO   
484615        M0007              Guanajuato                FOMIX GUANAJUATO   
831972        M0007              Guanajuato                FOMIX GUANAJUATO   
231236        M0007              Guanajuato                FOMIX GUANAJUATO   
663422        M0007              Guanajuato                FOMIX GUANAJUATO   
...             ...                     ...                             ...   
1976318       FOP07                    FPNE                        PRONACES   
1975070       FOP07                    FPNE                        PRONACES   
1975044       FOP10                    FVNI  Vinculación E Nacional Interna   
1951411       FOP10                    FVNI  Vinculación E Nacional Interna   
1976454       FOP10                    FVNI  Vinculación E Nacional Interna   

        CLAVE_CONVOCATORIA  DESCRIPCION_LARGA_CONVOCATORIA  \
835971        M0007-2011-2               GUANAJUATO 2011-2   
484615        M0007-2008-4               GUANAJUATO 2008-4   
831972        M0007-2011-1               GUANAJUATO 2011-1   
231236        M0007-2007-1               GUANAJUATO 2007-1   
663422        M0007-2009-3               GUANAJUATO 2009-3   
...                    ...                             ...   
1976318      FOP07-2022-01      SOBERANÍA ALIMENTARIA 2022   
1975070      FOP07-2021-04        SISTEMAS SOCIOECOLÓGICOS   
1975044      FOP10-2021-02  MEDICINA TRADICIONAL Y HERBOLA   
1951411      FOP10-2021-01              Asignación Directa   
1976454      FOP10-2022-01          Encargo de Estado 2022   

        FECHA_INICIO_CONVOCATORIA   FECHA_FIN_CONVOCATORIA CLAVE_MODALIDAD  \
835971    2006-06-07 00:00:00.000  2006-07-13 00:00:00.000              D1   
484615    2006-06-07 00:00:00.000  2006-07-13 00:00:00.000              A1   
831972    2006-06-07 00:00:00.000  2006-07-13 00:00:00.000              B1   
231236    2006-06-07 00:00:00.000  2006-07-13 00:00:00.000              A1   
663422    2006-06-07 00:00:00.000  2006-07-13 00:00:00.000              C1   
...                           ...                      ...             ...   
1976318   2022-02-18 00:00:00.000  2022-10-31 00:00:00.000              C3   
1975070   2022-02-18 00:00:00.000  2022-10-31 00:00:00.000              C1   
1975044   2022-02-28 00:00:00.000  2022-12-30 00:00:00.000              C1   
1951411   2022-02-28 00:00:00.000  2022-12-30 00:00:00.000              B2   
1976454   2022-02-28 00:00:00.000  2022-12-30 00:00:00.000             BB1   

                              MODALIDAD CLAVE_PROYECTO  
835971   Fortalecimiento Infraestructur         154549  
484615      Invest. Científica Aplicada         107682  
831972           Desarrollo Tecnológico         174020  
231236      Invest. Científica Aplicada          68602  
663422    Formación de Recursos Humanos         127629  
...                                 ...            ...  
1976318                             NaN         321299  
1975070   Formación de Recursos Humanos         319103  
1975044   Formación de Recursos Humanos         319042  
1951411                             NaN         316535  
1976454                             NaN         321321  

[1437 rows x 10 columns]